src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp

```diff
@@ -1895,6 +1849,202 @@ void mlir::checkConstantOutputs(
   }
 }
 
+void mlir::recordEntryPointSignatures(ModuleOp &module,
+    SmallVectorImpl<std::string> &entryPointNames,
+    SmallVectorImpl<std::string> &inSignatures,
+    SmallVectorImpl<std::string> &outSignatures) {
+  module->walk([&](KrnlEntryPointOp entryOp) -> WalkResult {
+    Operation *op = entryOp.getOperation();
+    // Entry point name.
+    llvm::StringRef entryPointName =
+        op->getAttrOfType<SymbolRefAttr>(
+              KrnlEntryPointOp::getEntryPointFuncAttrName())
+            .getLeafReference()
+            .getValue();
+    std::string terminatedEntryPointName = "run_" + entryPointName.str();
+    terminatedEntryPointName.push_back('\0'); // null terminate the string.
+    entryPointNames.emplace_back(terminatedEntryPointName);
+
+    // Input/output signatures.
+    StringAttr sigAttr =
+        op->getAttrOfType<StringAttr>(KrnlEntryPointOp::getSignatureAttrName());
+    llvm::StringRef signature = sigAttr.getValue();
+    auto splitSig = signature.split('@');
+    llvm::StringRef inSig = splitSig.first;
+    llvm::StringRef outSig = splitSig.second;
+    inSignatures.emplace_back(inSig.str());
+    outSignatures.emplace_back(outSig.str());
+
+    return WalkResult::advance();
+  });
+  // When there is only a single entry point function in a model, use
+  // DEFAULT_DYN_ENTRY_POINT.
+  if (entryPointNames.size() == 1) {
+    entryPointNames[0] = DEFAULT_DYN_ENTRY_POINT;
+    entryPointNames[0].push_back('\0'); // null terminate the string.
+  }
+}
+
+void mlir::genSignatureFunction(ModuleOp module,
+    ArrayRef<std::string> entryPointNames, ArrayRef<std::string> inSignatures,
```

Comment:
`entryPointNames`, `inSignatures ` and `outSignatures ` should not be modified by this function, suggest to make them const.